(RE-6640) Service Specific Files for Vanagon

jcoconnor opened the pull request:
This is the initial set of service specific files to support the Puppet
Agent Vanagon Windows build.

This PR replaces https://github.com/puppetlabs/puppet-agent/pull/514/

jcoconnor commented:
This is ready for merge (subject for review of course).
The service files compile, but as yet, the light.exe command doesn't produce and MSI due to some other missing parts.
Compile log: https://gist.github.com/jcoconnor/1810099576006228bd5b

I'm requesting that we merge these now to allow the Custom Action and UI fragment files to be added and to allow @McdonaldSeanp to proceed with a directory table to support these service files.

/cc @melissa @jpogran @Iristyle @ferventcoder

melissa commented on configs/components/pxp-agent.rb:
```diff
@@ -103,7 +103,9 @@
-    puts "Service files not enabled on windows"
+    # Note - this definition indicates that the file should be filtered out from the Wix
+    # harvest. A corresponding service definition file is also required in resources/windows/wix
+    pkg.install_service "SourceDir\\ProgramFiles64Folder\\PuppetLabs\\Puppet\\bin\\nssm.exe"
```
should we make this a variable so people don't have to type it twice?

McdonaldSeanp replied:
@melissa are you talking about the directory structure above the .exe s? The precedent set by the other platforms is to leave the entire path in. unless we wanted to change all I would say no variable

melissa replied:
We don't have any variables that help with the other paths, since they're getting pulled from the local repos. Since we're installing `nssm.exe`, we know where it's going to go already (https://github.com/puppetlabs/puppet-agent/blob/master/configs%2Fcomponents%2Fnssm.rb#L11). In this case, we're installing nssm to bindir. Bindir currently evaluates to `C:/ProgramFiles64Folder/PuppetLabs/Puppet/puppet/bin`, so with a little path munging, if we ever change the location of nssm.exe, then we'd only have to change it in one place rather than three.

melissa replied:
but the path munging required might make that not worth it. I dunno, I'll leave it up to you

jcoconnor replied:
Initially I thought no to using variables, but since the service files are .erb files, no reason why we can't use a variable here. Will try it out tammara

jcoconnor replied:
I have decided not to use a variable for the moment - mainly due to the \\ and \ quoting rules.

melissa commented on resources/windows/wix/service.mcollective.wxs.erb:
```diff
@@ -0,0 +1,28 @@
+          Type="ownProcess"
+          ErrorControl="normal"
+          Vital="yes"
+          Arguments="-I&quot;[INSTALLDIR]puppet\lib;[INSTALLDIR]facter\lib;[INSTALLDIR]hiera\lib;[INSTALLDIR]mcollective\lib&quot; -rubygems &quot;[INSTALLDIR]mcollective\bin\mcollectived&quot; --daemonize"
```
This arguments are no longer correct with the new libdir layout

jcoconnor replied:
Agreed - Think they are correct now - please verify

melissa commented on resources/windows/wix/service.mcollective.wxs.erb:
```diff
@@ -0,0 +1,28 @@
+<?xml version='1.0' encoding='windows-1252'?>
+<Wix xmlns='http://schemas.microsoft.com/wix/2006/wi' xmlns:util="http://schemas.microsoft.com/wix/UtilExtension">
+  <Fragment>
+    <DirectoryRef Id="PUPPETBINDIR">
```
Did you get this working? Where are we defining this directory?

McdonaldSeanp replied:
We do not have this working, However at this point I think it will be better to merge this PR with this in to facilitate other work. Eventually this will either be searched for or defined in the directorylist.wxs file I am working on. EIther way this will stay in. What john was proposing (an idea that I like) is that for now we can merge wix files that can pass the candle step. Then when we have moved past getting light to execute we can make sure all directoryrefs work.

melissa commented on resources/windows/wix/service.pxp-agent.wxs.erb:
```diff
@@ -0,0 +1,45 @@
+    <DirectoryRef Id="PUPPETBINDIR">
+      <Component Id="service_nssm" Guid="52B1CD57-95A2-4CA4-AB8E-9DDD6DE8FC66">
+        <CreateFolder />
+        <File Id="NSSM" KeyPath="yes" Source="SourceDir\ProgramFiles64Folder\PuppetLabs\Puppet\bin\nssm.exe" />
```
This should be `SourceDir\ProgramFiles64Folder\PuppetLabs\Puppet\puppet\bin\nssm.exe` as per https://github.com/puppetlabs/puppet-specifications/pull/63

jcoconnor replied:
Corrected - see note below

melissa commented on resources/windows/wix/service.pxp-agent.wxs.erb:
```diff
@@ -0,0 +1,45 @@
+        <!-- Various registry keys documented at https://nssm.cc/usage -->
+        <RegistryKey Root="HKLM" Key="SYSTEM\CurrentControlSet\Services\pxp-agent\Parameters">
+          <RegistryValue Name="AppDirectory" Value="[INSTALLDIR]bin" Type="expandable" />
+          <RegistryValue Name="Application" Value="[INSTALLDIR]bin\pxp-agent.exe" Type="expandable" />
```
`Value="[INSTALLDIR]\puppet\bin\pxp-agent.exe"`

jcoconnor replied:
Corrected - it should be ```Value="[INSTALLDIR]puppet\bin\pxp-agent.exe"```

melissa commented:
@jcoconnor a lot of these paths are wrong

jcoconnor commented:
@melissa Wasn't sure about the paths - esp the nssm ones. Why are we putting it down a level (puppet) for nssm.exe but not ruby.exe?

melissa commented:
@jcoconnor ruby is on the same path as nssm, and is wrong in this PR

jcoconnor commented:
@melissa and @McdonaldSeanp 
1. I have gone thru the service files and *.rb scripts and corrected the pathnames.
2. I have also updated the test build with the latest puppet-agent....tar.gz which has a number of pathname fixes - this now appears to match the [file specification PR](https://github.com/puppetlabs/puppet-specifications/pull/63/files)

Can you check the arguments for the mco and puppet-agent service files - I think I have specified the -rubygems correctly, but would appreciate if you could verify.

melissa commented on configs/components/marionette-collective.rb:
```diff
@@ -66,7 +66,9 @@
-    puts "Service files not enabled on windows"
+    # Note - this definition indicates that the file should be filtered out from the Wix
+    # harvest. A corresponding service definition file is also required in resources/windows/wix
+    pkg.install_service "SourceDir\\ProgramFiles64Folder\\PuppetLabs\\Puppet\\puppet\\bin\\rubyw.exe"
```
It's not always going to be `ProgramFiles64Folder`. https://github.com/puppetlabs/puppet-agent/blob/master/configs%2Fprojects%2Fpuppet-agent.rb#L19

jcoconnor replied:
So can I just put in ```#{settings[:base_dir]}``` instead?

McdonaldSeanp replied:
Yeah, I would do that

jcoconnor replied:
Thats done - Have also rebased the changes to master again.

melissa commented on configs/projects/puppet-agent.rb:
```diff
@@ -6,6 +6,7 @@
     proj.setting(:company_name, "Puppet Labs")
     proj.setting(:company_id, "PuppetLabs")
     proj.setting(:common_product_id, "PuppetInstaller")
+    proj.setting(:PuppetServiceName, "puppet")
```
Rather than using camel case here, I'd prefer we keep it in the ruby convention with underscores. Not a blocker, just a personal preference. If that opinion isn't shared, I'm not going to fight for it.

McdonaldSeanp replied:
agreed, underscores

jcoconnor replied:
You mean like ```puppet_service_name``` ?
Thats fine - I need to fix the program files folder anyway.

jcoconnor replied:
Done

melissa commented on resources/windows/wix/service.mcollective.wxs.erb:
```diff
@@ -0,0 +1,28 @@
+  <Fragment>
+    <DirectoryRef Id="PUPPETBINDIR">
+      <Component Id='MCOService' Guid="7601FCEA-90B3-CC69-6A69-4087FBC7292D" Win64="<%= settings[:win64] %>">
+        <File Id="RubyWExe" KeyPath="yes" Source="SourceDir\ProgramFiles64Folder\PuppetLabs\Puppet\puppet\bin\rubyw.exe" />
```
`ProgramFiles64Folder` needs to change here too

jcoconnor replied:
Done

melissa commented on resources/windows/wix/service.puppet.wxs.erb:
```diff
@@ -0,0 +1,73 @@
+          Description="Periodically fetches and applies configurations from a Puppet master server."
+          DisplayName="Puppet Agent"
+          Interactive="no"
+          Name="<%= settings[:PuppetServiceName] %>"
```
puppet_service_name

jcoconnor replied:
Yeah - just spotted that in test :(  Fixed and pushed up.

melissa commented on resources/windows/wix/service.pxp-agent.wxs.erb:
```diff
@@ -0,0 +1,45 @@
+    <DirectoryRef Id="PUPPETBINDIR">
+      <Component Id="service_nssm" Guid="52B1CD57-95A2-4CA4-AB8E-9DDD6DE8FC66">
+        <CreateFolder />
+        <File Id="NSSM" KeyPath="yes" Source="SourceDir\<%= settings[:base_dir] %>\PuppetLabs\Puppet\puppet\bin\nssm.exe" />
```
`PuppetLabs` and `Puppet` have corresponding settings, and could be parameterized

jcoconnor replied:
Are they already used to define directories?

melissa replied:
https://github.com/puppetlabs/puppet-agent/blob/master/configs%2Fprojects%2Fpuppet-agent.rb#L6-L9

jcoconnor replied:
OK. Will sort Mon.

jcoconnor replied:
Done

jcoconnor commented:
@jpogran Can you please review this today ?
Also - /CC @Iristyle @ferventcoder @melissa

jcoconnor commented:
I have just pushed some changes to this PR in connection with Vanagon PR: https://github.com/puppetlabs/vanagon/pull/308

melissa commented on resources/windows/wix/service.mcollective.wxs.erb:
```diff
@@ -0,0 +1,40 @@
+<?xml version='1.0' encoding='windows-1252'?>
+<Wix xmlns='http://schemas.microsoft.com/wix/2006/wi' xmlns:util="http://schemas.microsoft.com/wix/UtilExtension">
+  <Fragment>
+    <ComponentGroup Id="Service_MARIONETTECOLLECTIVE">
```
Does this correspond to `:id => "#{service_name.gsub(/([^A-Za-z0-9])/, '').upcase}"`? Where is `:id` used?

jcoconnor replied:
@melissa Yes - thats where :id is used.
I was going to change :id to Service_(gsubbed stuff), but after looking at Sean's code, he can also make use of the :id if we leave it without a Service_ prefix.

melissa replied:
If that's the case, can we make this a variable? Especially since `id` has some syntax requirements

jcoconnor replied:
OK - thats another MAINT Fix that I need to push up on the Vanagon Side - will be having BJ with @McdonaldSeanp shortly so will discuss with him as it impacts his mods too.

melissa replied:
Okay, I'm seeing some discrepancies. In https://github.com/puppetlabs/vanagon/commit/77664321791beeb01d29ebdf3cbb941a78b669e0, we set `:id => "#{service_name.gsub(/([^A-Za-z0-9])/, '').upcase}"` but then when we reference that in the wix files, we prepend `Service_` to it. Which is it? `MARIONETTECOLLECTIVE` or `Service_MARIONETTECOLLECTIVE`? Are we ever going to need `service.id` to be `MARIONETTECOLLECTIVE` instead of `Service_MARIONETTECOLLECTIVE`?

McdonaldSeanp replied:
So here are the things at play here: 

* I need to get :id as it is now (without the prepended service_) for the directorylist work, so when we are creating the id field in the component dsl, it should not have Service_

* The point of prepending Service_ serves as a measure of documentation. Specifically for me the prepended service makes it clear that this component element contains the service WiX for MCO and not any of the contents of MCO itself. So in this way I like that the component element has Id="Service_MARIONETTECOLLECTIVE" and not just Id="MARIONETTECOLLECTIVE"

McdonaldSeanp replied:
Also We can't really make this a variable, since when vanagon would get to this erb it would not have context as to which service it was editing this file for, so unless we did something like take a hard line on the filename and say the filename needs to be the name of the service there is no way for us to know which service to put here.

jcoconnor commented:
For reference - this is now building against Sean's latest DirList changes in Vanagon: https://gist.github.com/jcoconnor/0b15bfaf6b4fd4144a37

glennsarti commented:
:+1: from me.  Merge it.

melissa commented on resources/windows/wix/service.mcollective.wxs.erb:
```diff
@@ -0,0 +1,40 @@
+      <ComponentRef Id="MCOService" />
+    </ComponentGroup>
+
+    <DirectoryRef Id="PUPPETBINDIR">
```
Should this be `PUPPETBINDIR` or `MARIONETTECOLLECTIVEBINDIR`?

McdonaldSeanp replied:
It should be MARIONETTECOLLECTIVEBINDIR

melissa commented on resources/windows/wix/service.pxp-agent.wxs.erb:
```diff
@@ -0,0 +1,61 @@
+
+  <Fragment>
+    <ComponentGroup Id="Service_PXPAGENT">
+      <ComponentRef Id="Service_NSSM" />
```
Can we have this be consistent? I know it's isolated to the service itself, but we currently have three different ways we're defining this. `<ComponentRef Id="MCOService" />`, <ComponentRef Id="PuppetServiceAutomatic" /> (and friends), and `<ComponentRef Id="Service_NSSM" />`.

melissa replied:
Or, wait, do these need to be consistent from pftw? If that's the case, then ignore that last comment.

jcoconnor replied:
They don't have to be consistent with pftw - I think I was just using the original ID's.
Will modify it slightly.
You are correct about the BINDIRs - will fix these now too.

jcoconnor replied:
Have just pushed up an additional commit for these.
Will squash the commits once I have tested a local build.

melissa commented on resources/windows/wix/service.pxp-agent.wxs.erb:
```diff
@@ -0,0 +1,61 @@
+      <ComponentRef Id="Service_NSSM" />
+    </ComponentGroup>
+
+    <DirectoryRef Id="PUPPETBINDIR">
```
Should this be `PXPAGENTBINDIR`?

melissa commented on resources/windows/wix/service.mcollective.wxs.erb:
```diff
@@ -0,0 +1,40 @@
+    </ComponentGroup>
+
+    <DirectoryRef Id="MARIONETTECOLLECTIVEBINDIR">
+      <Component Id='Service_MARIONETTECOLLECTIVE_svc'
```
I mean, I kind of preferred the `MCOService` syntax, since that means we don't have to spell out MARIONETTECOLLECTIVE everytime in capital letters.... also, if you do want to stick with this syntax, can we please please please please use the mco service ID that gets set already?

jcoconnor replied:
I guess one way we could do this simply (lets MAINT it though) is to add a serviceref or shortref key in the ```install_service``` method, where we would use this instead of the full ```MARIONCOLLECTIVE``` name. (just thinking on the hoof). If this is not provided, then it would default to the full length name.

McdonaldSeanp replied:
@melissa That's fine, I can open up a maint PR that overrides the service name to MCO. Which will make all these become MCO instead of MARIONETTECOLLECTIVE. And @jcoconnor overriding the name in install service is already provided. If you pass in install service like so: `install_service #{path_to_bin} nil #{name}` it will actually already do the name override for us =) no need to add anything to install_service. refer to: https://github.com/McdonaldSeanp/vanagon/blob/feature/RE-6455/dirlist/spec/lib/vanagon/platform/windows_spec.rb#L186L197
